larq_compute_engine/mlir/transforms/optimize.cc

```diff
@@ -130,23 +213,117 @@ struct SetBitpackedActivations : public OpRewritePattern<BinaryOp> {
     // Try and match `input_op` to a binary convolution.
     auto inner_bconv_op = dyn_cast_or_null<TF::LceBconv2dOp>(input_op);
     if (inner_bconv_op) {
-      if (inner_bconv_op.padding() == "SAME" &&
-          inner_bconv_op.pad_values() != 1) {
+      // We can't apply this transform if the activation or filters are already
+      // bitpacked, or with 'same zero' padding.
+      const auto inner_type = inner_bconv_op.getType().cast<ShapedType>();
+      const auto filter_type =
+          inner_bconv_op.filter().getType().cast<ShapedType>();
+      if (!inner_type.getElementType().isF32() ||
+          !filter_type.getElementType().isF32() ||
+          (inner_bconv_op.padding() == "SAME" &&
+           inner_bconv_op.pad_values() != 1)) {
         return failure();
       }
 
       if (auto maybe_bitpacked_type = maybeGetBitpackedType(
               rewriter, inner_bconv_op.getType().cast<ShapedType>())) {
+        // As the inner bconv op will be writing bitpacked output, we need to
+        // compute the thresholds for writing a 1-bit or a 0-bit.
+
+        // Compute the constants `backtransform_add` and `clamp_min/max`.
+        const auto filter_shape = filter_type.getShape();
+        const std::int32_t backtransform_add =
+            filter_shape[1] * filter_shape[2] * filter_shape[3];
+        const auto clamp_min_max =
+            llvm::StringSwitch<std::pair<std::int32_t, std::int32_t>>(
+                inner_bconv_op.fused_activation_function())
+                .Case("RELU", {0, backtransform_add})
+                .Case("RELU_N1_TO_1", {-1, 1})
+                .Case("RELU6", {0, 6})
+                .Default({-1 * backtransform_add, backtransform_add});
```

Comment:
I've used this rather than the minimum/maximum representable Int32 value because I was for some reason getting some overflow errors and ended up with `clamp_min = clamp_max = min_int32`. This is valid though, as the xor-popcount accumulator will never be outside this range.